File: android/src/main/kotlin/com/mapbox/maps/mapbox_maps/snapshot/SnapshotController.kt
package com.mapbox.maps.mapbox_maps.snapshot

import android.content.Context
import android.os.Looper
import com.mapbox.maps.MapSnapshotOptions
import com.mapbox.maps.MapView
import com.mapbox.maps.Size
import com.mapbox.maps.SnapshotOverlayOptions
import com.mapbox.maps.SnapshotStyleListener
import com.mapbox.maps.Snapshotter
import com.mapbox.maps.Style
import com.mapbox.maps.mapbox_maps.pigeons.*
import com.mapbox.maps.mapbox_maps.toMbxImage
import io.flutter.plugin.common.BinaryMessenger
import kotlinx.coroutines.runBlocking
import java.util.UUID
import java.util.logging.Handler

interface SnapshotControllerDelegate {
  fun getSnapshotter(id: String): Snapshotter

  fun getContext(): Context

  fun remove(id: String)
}

class SnapshotController(private val mapView: MapView, private val context: Context) :
  _SnapShotManager,
  SnapshotControllerDelegate {

  private val snapshotterMap = mutableMapOf<String, Snapshotter>()
  private lateinit var onSnapshotStyleListener: OnSnapshotStyleListener
  private val snapshotter: SnapshotterManager = SnapshotterManager(this)

  override fun create(
    options: com.mapbox.maps.mapbox_maps.pigeons.MapSnapshotOptions,
    overlayOptions: com.mapbox.maps.mapbox_maps.pigeons.SnapshotOverlayOptions
  ): String {
    val snapshotter =
      Snapshotter(
        context,
        options.toSnapshotOptions(),
        overlayOptions.toSnapshotOverlayOptions()
      ).apply {
        this.setStyleListener(object : SnapshotStyleListener {
          override fun onDidFinishLoadingStyle(style: Style) {
            onSnapshotStyleListener.onDidFinishLoadingStyle {}
          }

          override fun onDidFailLoadingStyle(message: String) {
            onSnapshotStyleListener.onDidFailLoadingStyle(message) {}
          }

          override fun onDidFullyLoadStyle(style: Style) {
            onSnapshotStyleListener.onDidFullyLoadStyle({})
          }

          override fun onStyleImageMissing(imageId: String) {
            onSnapshotStyleListener.onStyleImageMissing(imageId) {}
          }
        })
      }
    val id = UUID.randomUUID().toString().replace("-", "")
    snapshotterMap[id] = snapshotter
    return id
  }


  private var runnable: Runnable? = null
  private val handler = android.os.Handler(Looper.getMainLooper())

  override fun snapshot(callback: (Result<MbxImage?>) -> Unit) {
    var image: MbxImage? = null
    runnable = Runnable {
      callback(Result.success(image))
      runnable = null
    }
    try {
      mapView.javaClass.classLoader?.loadClass("com.mapbox.maps.renderer.gl.PixelReader")?.let {clz->
        val field = clz.getDeclaredField("supportsPbo")
        field.isAccessible = true
        field.set(null, false)
      }
      mapView.snapshot {
        image = it?.toMbxImage()
        runnable?.apply { handler.postDelayed(this, 500) }
      }
    } catch (t:Throwable){
      mapView.snapshot {
        image = it?.toMbxImage()
        runnable?.apply { handler.postDelayed(this, 500) }
      }
    }
  }

  private fun com.mapbox.maps.mapbox_maps.pigeons.MapSnapshotOptions.toSnapshotOptions(): MapSnapshotOptions {
    val options = MapSnapshotOptions.Builder()
    options.size(Size(this.size.width.toFloat(), this.size.height.toFloat()))
    options.pixelRatio(this.pixelRatio.toFloat())
    return options.build()
  }

  private fun com.mapbox.maps.mapbox_maps.pigeons.SnapshotOverlayOptions.toSnapshotOverlayOptions(): SnapshotOverlayOptions {
    return SnapshotOverlayOptions(this.showLogo, this.showAttributes)
  }

  override fun getSnapshotter(id: String): Snapshotter {
    if (snapshotterMap[id] == null) {
      throw (Throwable("No Snapshotter found with id: $id"))
    }
    return snapshotterMap[id]!!
  }

  override fun getContext(): Context {
    return context
  }

  override fun remove(id: String) {
    snapshotterMap.remove(id)
  }

  fun setup(messenger: BinaryMessenger) {
    onSnapshotStyleListener = OnSnapshotStyleListener(messenger)
    _SnapShotManager.setUp(messenger, this)
    _SnapshotterMessager.setUp(messenger, snapshotter)
  }

  fun dispose(messenger: BinaryMessenger) {
    _SnapShotManager.setUp(messenger, null)
    _SnapshotterMessager.setUp(messenger, null)
    snapshotterMap.clear()
  }
}

File: ios/Classes/SnapshotController.swift
import Foundation
import MapboxMaps
import Flutter

public enum SnapshotControllerError: Error {
    case noSnapshotterFound
}

public protocol SnapshotControllerDelegate : AnyObject {
    func getSnapshotter(id: String) throws -> Snapshotter
    
    func remove(id: String)
}

class SnapshotController: NSObject,_SnapShotManager,SnapshotControllerDelegate {
 
    private var mapView: MapView
    private var onSnapshotStyleListener: OnSnapshotStyleListener?
    private var snapshotterMap: [String: Snapshotter] = [:]
    private var cancelables: [String: Set<AnyCancelable>] = [:]
    private var snapshotter: SnapshotterManager?
    
    init(mapView: MapView) {
        self.mapView = mapView
        super.init()
        snapshotter = SnapshotterManager(withDelegate: self)
    }
    
    func create(options: MapSnapshotOptions, overlayOptions: SnapshotOverlayOptions) throws -> String {
        let snapshotOptions = MapboxMaps.MapSnapshotOptions.init(size: CGSize(width: options.size.width, height: options.size.height), pixelRatio: options.pixelRatio,showsLogo: overlayOptions.showLogo,showsAttribution: overlayOptions.showAttributes)
        let snapshotter = Snapshotter(options: snapshotOptions)
        let id = UUID().uuidString.replacingOccurrences(of: "-", with: "")
        snapshotterMap[id] = snapshotter
        
        cancelables[id] = Set()
        
        snapshotter.onStyleLoaded.observe { [weak self] style in
            self?.onSnapshotStyleListener?.onDidFinishLoadingStyle(completion: { flutterError in
            })
        }.store(in: &cancelables[id]!)
    
        snapshotter.onStyleDataLoaded.observe { [weak self] styleDataLoad in
            self?.onSnapshotStyleListener?.onDidFullyLoadStyle(completion: { flutterError in
                
            })
        }.store(in: &cancelables[id]!)
        
        snapshotter.onMapLoadingError.observe { [weak self] mapLoadingError in
            self?.onSnapshotStyleListener?.onDidFailLoadingStyle(message: mapLoadingError.message, completion: { flutterError in
            })
        }.store(in: &cancelables[id]!)
        
        snapshotter.onStyleImageMissing.observe { [weak self] styleImageMissing in
            self?.onSnapshotStyleListener?.onStyleImageMissing(imageId: styleImageMissing.imageId, completion: { flutterError in
            })
        }.store(in: &cancelables[id]!)
        
        return id
    }
    
    func snapshot(completion: @escaping (Result<MbxImage?, Error>) -> Void) {
        let uiImage = try? mapView.snapshot()
        guard let image = uiImage else {
            completion(Result.success(nil))
            return
        }
        completion(Result.success(image.toFLTMbxImage()))
    }
    
    func setup(messager: FlutterBinaryMessenger) {
        _SnapShotManagerSetup.setUp(binaryMessenger: messager, api: self)
        onSnapshotStyleListener = OnSnapshotStyleListener.init(binaryMessenger: messager)
        _SnapshotterMessagerSetup.setUp(binaryMessenger: messager, api: snapshotter)
    }
    
    
    func getSnapshotter(id: String) throws -> MapboxMaps.Snapshotter {
        if snapshotterMap[id] == nil {
            throw SnapshotControllerError.noSnapshotterFound
        }
        return snapshotterMap[id]!
    }
    
    func remove(id: String) {
        snapshotterMap.removeValue(forKey: id)
        cancelables[id]?.forEach({ anyCancelable in
            anyCancelable.cancel()
        })
        cancelables[id]?.removeAll()
    }

    func tearDown(messenger: FlutterBinaryMessenger) {
        _SnapShotManagerSetup.setUp(binaryMessenger: messenger, api: nil)
        _SnapshotterMessagerSetup.setUp(binaryMessenger: messenger, api: nil)
        onSnapshotStyleListener = nil
        cancelables.keys.forEach { key in
            remove(id: key)
        }
        snapshotterMap.removeAll()
        cancelables.removeAll()
        snapshotter = nil
    }
}
